### ISSUE TITLE
Dracut should include `/etc/crypttab` when including `cryptsetup` and `systemd`

### ISSUE THREAD

**Issue by JustArchi** [dracutdevs/dracut#1737]
# Context

`systemd` uses `/etc/crypttab` file as a way to decrypt LUKS volumes before proceeding to `/etc/fstab` and mounting the partitions, including those that might be hidden behind it.

I'm testing standard gentoo installation with `systemd` and `gentoo-kernel-bin`, which uses `dracut` for initramfs. After embedding `cryptsetup` module, boot process is stuck - correctly, because nobody told embedded `systemd` what to do.

One solution to the problem, in my opinion vastly inferior, is specifying `rd.luks.uuid=` in the command-line, which will decrypt that one LUKS partition. It gets more complicated when we'd like to decrypt more than one partition, for example two on two different drives. I didn't have any luck achieving satisfying setup this way.

However, what I did was creating a very simple `/etc/crypttab` file which holds a recipe for decryption:

```
sda3_crypt UUID=50d2da25-xxxx-xxxx-xxxx-xxxxxxxxxxxx none luks,discard
sdb1_crypt UUID=1cb94969-yyyy-yyyy-yyyy-yyyyyyyyyyyy none luks,discard
```

I was surprised that dracut didn't embed this file on its own. I required only one last bit in the puzzle, actually embedding this file in the final dracut initramfs:

```
install_items+=" /etc/crypttab "
```

After reboot, I was greeted to password prompt, moreover, utilizing systemd functionality of re-using the password for other, entirely independent LUKS partitions, in result I was asked for password only once. I also didn't have any errors in the log or similar, which I did have with `rd.luks.uuid=` as systemd effectively tried to decrypt the already decrypted and mounted device twice, once it finally got to the real `/etc/crypttab`.

# Feature request

What I'd like to see is this one, very simple line which gets added by default when dracut is configured to use `systemd` and `cryptsetup`:

```
install_optional_items+=" /etc/crypttab "
```

That is, if `/etc/crypttab` exists with `systemd` and `cryptsetup` modules enabled, embed it in initramfs and allow `systemd` to do its decryption thing right, without any need of additional grub cmdline args or user in need to configure it manually. This is what happens e.g. in Debian - it doesn't use dracut, but it does embed `/etc/crypttab` in its systemd initramfs, allowing the boot process to smoothly decrypt LUKS partitions without additional hassle.

Now I'm not sure if the proper repo for this issue is dracut or gentoo directly, but I believe it should aid more distros than gentoo alone, and I really don't see a good reason for this behaviour to not be a default one - we don't require that file explicitly, we just include it for systemd because we can be 99.9% sure that if user has `systemd`, `cryptsetup`, and `/etc/crypttab`, then it's the correct thing to do. Worst case user finds out his `/etc/crypttab` doesn't make sense, or decide to not include it in dracut because he knows better. It should be treated same as `/etc/fstab` inclusion in this regard.

Thank you in advance for considering this enhancement.

**Comment by tacerus** [comment]
The crypt modules contains a cmdline function which automatically appends `rd.luks.uuid=` upon building the initramfs: https://github.com/dracutdevs/dracut/blob/aebeb2ecdf76b7716fadbab8b216b8805d36a461/modules.d/90crypt/module-setup.sh#L69 - Did it not contain this additional cmdline in the dracut output upon building the initramfs?

**Comment by JustArchi** [comment]
I'm not an expert, I did `lsinitrd` on generated initramfs and there is `/etc/cmdline.d/90crypt.conf` file with ` rd.luks.uuid=luks-50d2da25-xxxx-xxxx-xxxx-xxxxxxxxxxxx` content, so it looks like the logic has triggered properly.

However, despite of that it didn't work for me, the boot process was stuck until it timed out and returned to emergency shell. I don't know why.

However, even if it did work properly, there is still added value from relying on systemd to decrypt luks, as due to the fact that I use the same password for both luks partitions, I'm only asked for them once.

**Comment by tacerus** [comment]
That looks good. The behavior with the cmdline parameter is the same for both systemd based and traditional LUKS. It should build `/etc/crypttab` automatically upon building the initramfs and use systemd if it's available. If you are certain `/etc/crypttab` did not show up in `lsinitrd` before you added it manually, then I'm unfortunately out of ideas.

**Comment by JustArchi** [comment]
I did some further digging and think I found a reason, `/etc/crypttab` is not included automatically unless `hostonly="yes"` is used. This makes... total sense actually, but I had no clue previously that it was the case, as I started using `hostonly` mode only after opening this issue.

Moreover, it looks like even my argument about one-password decryption isn't valid, because it seems that despite `hostonly` option including only one line of `/etc/crypttab` (the rootfs), `systemd` is able to pick up real `/etc/crypttab` from decrypted rootfs and re-use the password for additional partitions without prompting for it.

I don't believe there is anything to improve in dracut logic in regards to this issue, perhaps the documentation should just be more explicit in this regard:

```
   Boot parameters
       An initramfs generated without the "hostonly" mode, does not
       contain any system configuration files (except for some special
       exceptions), so the configuration has to be done on the kernel
       command line.
(...)
```

Maybe something like this would be better:

```
   Boot parameters
       An initramfs generated without the "hostonly" mode, does not
       contain any system configuration files (except for some special
       exceptions), so the configuration has to be done on the kernel
       command line. This might require from you explicitly passing
       additional information (for example cryptsetup root), as opposed
       to being generated automatically when hostonly mode is enabled.
(...)
```

In any case, thank you for help @tacerus, as I'm happy to close this issue as no longer relevant.

**Comment by tacerus** [comment]
Oh right, I should have mentioned the host-only dependency.

Glad you were able to figure it out! :-)

**Comment by HugKitten** [comment]
Mkinitcpio also checks for files in /etc/crypttab.initramfs, so perhaps Dracut should look for if the file exist in /etc/crypttab.initramfs before using /etc/crypttab.
